Add Solana Payment Processing Implementation

racket2000 opened the pull request:
*Issue #, if available:* N/A

*Description of changes:*

This PR adds a complete Solana implementation of the serverless digital asset payment system, following the same architecture as the EVM version but adapted for Solana's unique characteristics.

### Overview
Implements invoice generation, payment monitoring, and automated fund sweeping for both native SOL and SPL tokens on Solana devnet/mainnet. The solution inherits key security features and design considerations from the EVM-compatible design. 

### Key Features
- **Native SOL Payments** - Direct SOL transfers with automatic detection and sweeping
- **SPL Token Support** - Full support for any SPL token (USDC, BONK, etc.) with dynamic decimal handling
- **HD Wallet Derivation** - BIP44 path m/44'/501'/x'/0' for unique invoice addresses
- **Automated Payment Flow** - Watcher Lambda detects payments, Sweeper Lambda transfers to treasury
- **API Gateway Integration** - RESTful API for invoice creation and management
- **QR Code Generation** - Automatic QR codes for easy mobile payments

### Architecture Differences from EVM
- Uses Solana JSON RPC instead of Ethereum JSON RPC
- Ed25519 key derivation instead of secp256k1
- SPL tokens instead of ERC20
- Associated Token Accounts for SPL token handling
- Lamports for fees instead of gas in Wei
- Hot Wallet Security: Uses AWS KMS with ed25519 signing.

### Technical Highlights
- **Dynamic Decimal Handling** - Automatically fetches token decimals from mint (supports any SPL token)
- **Efficient Sweeping** - DynamoDB Streams trigger fund transfers to treasury once payment is detected
- **Security Best Practices** - Secrets Manager for key storage, API key authentication, CDK Nag compliance
- **Automated Testing** - Test scripts for complete payment flow validation

### Testing -- The following have been validated 
-  SOL payment flow (create → pay → detect → sweep)
-  SPL token payment flow with USDC (6 decimals)
-  Dynamic decimal handling verified
-  All Lambda functions deployed and tested
-  CDK Nag security checks passing

### Documentation
- Complete independent Solana README with setup instructions, with hyperlink from main README
- API reference documentation
- Troubleshooting guide
- Automated test script examples

#### Files Added under [non-evm-deployments/Solana directory](https://github.com/aws-samples/sample-serverless-digital-asset-payments/tree/solana-blueprint)

Contains:
- CDK infrastructure (sonana/lib/solana-invoice-stack.ts)
-  Comprehensive documentation
- Refactored components include:
1. Lambda functions (invoice, watcher, sweeper, management).
2. Refactored payment scripts with .env integration
3. Refactored automated test scripts

Note: If you conduct a diff between the EVM and Solana components, you will observe that the changes to the Generator, Watcher, and Sweeper are quite minimal. 

This PR is ready for review and testing on Solana devnet. I nominate @dpdornseifer to complete this review, as he has a great amount of knowledge and experience regarding the EVM blueprint. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
Add Solana payment processing implementation

Commit:
remove USDC address from faucet instructions

Commit:
Incorporporated e2e payment testing for SOL and SPL invoices.

Commit:
polish

Commit:
add cd command to instructions

Commit:
add --require-approval never to npm run setup to automate setup process

Commit:
add bootstrap to quickstart, and fix bug in setup-secrets

Commit:
add bootstrap to quickstart, and fix bug in setup-secrets

Commit:
add error messages to setup-secrets

Commit:
add error handling to keys setup script

Commit:
Fix lint errors (remove unused imports)

Commit:
remove seed phrase from console when setting up secrets

Commit:
fix formatting w prettier

Commit:
repo restructure

Commit:
fix styling w prettier

Commit:
Improve deployment workflow and reduce output redundancy

- Add automated setup.sh script for end-to-end deployment
- Simplify Quick Start to use 'npm run setup' command
- Remove redundant funding instructions from generate-wallets.js
- Remove unnecessary treasury private key warning
- Add note about using custom wallets in README
- Fix table of contents to match actual sections
- Add navigation from Quick Start to Usage section

Commit:
Format README with prettier

Commit:
remove extraneous comment in sweeper and fix cd command

Commit:
Add optional SNS email subscription to setup script

- Prompt user to subscribe to payment notifications during setup
- Automatically retrieve SNS topic ARN from stack outputs
- Send subscription confirmation email via AWS SNS

Commit:
Format wallet-info.js with prettier

Commit:
Clarify payment methods in manual payment section

Commit:
change USDC invoice to be a smaller amount, since faucet only drips 1 usdc

Commit:
simply clean up command

Commit:
fix formatting

Commit:
Integrate AWS KMS Ed25519 for Solana hot wallet signing

- Replace Secrets Manager hot wallet private key with KMS Ed25519 key
- Update sweeper Lambda to use KMS signing API with ED25519_SHA_512
- Modify setup scripts to generate KMS key and derive hot wallet address
- Update wallet-info.js to use AWS SDK v3 (no deprecation warnings)
- Improve setup script output with clear wallet funding instructions
- Remove hot wallet private key from .env-sample
- Tested with both SOL and USDC payments successfully

Commit:
fix: resolve lint errors - add curly braces and remove unused variable

Commit:
nit pick styling

Commit:
fix decimal forrmatting bug in logs

racket2000 commented:
Completed:  Hot Wallet Private Key stored in KMS with ed25519 signing.

dpdornseifer commented on non-evm-deployments/solana/lambda/sweeper/index.js:
```diff
@@ -0,0 +1,269 @@
+
+async function sendErrorNotification(error, invoiceId) {
+  try {
+    const logGroupName = '/aws/lambda/SolanaInvoiceStack-SolanaSweeperFunction';
```
can we make the log group name variable and derive from stack name

racket2000 replied:
yes, fixed (this will have to be fixed in the EVM stack as well)

dpdornseifer commented on non-evm-deployments/solana/lib/solana-invoice-stack.ts:
```diff
@@ -0,0 +1,392 @@
+      secretObjectValue: {},
+    });
+
+    const hotWalletKmsKeyCfn = new cdk.aws_kms.CfnKey(this, 'SolanaHotWalletKey', {
```
per default KMS keys have `RemovalPolicy.RETAIN` - explicitly set to `DESTROY` to clean up all resources

racket2000 replied:
fixed

dpdornseifer commented on non-evm-deployments/solana/bin/solana-invoice.ts:
```diff
@@ -0,0 +1,9 @@
+import { SolanaInvoiceStack } from '../lib/solana-invoice-stack';
+
+const app = new cdk.App();
+// cdk.Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
```
cdk NAG should be turned on

racket2000 replied:
oops, I turned this off during testing and thought I had turned it back on. Thanks for calling this out.

racket2000 replied:
fixed

dpdornseifer commented on non-evm-deployments/solana/README.md:
````diff
@@ -0,0 +1,271 @@
+5. **Account Model**: Solana's account-based model with Associated Token
+   Accounts
+6. **Hot Wallet Security**: Uses AWS KMS with ed25519 signing
+
+## Prerequisites
+
+1. AWS Account
+2. Node.js 18.x or later
+3. AWS CDK CLI installed (`npm install -g aws-cdk`)
+
+## Quick Start
+
+```bash
+cd ~/sample-serverless-digital-asset-payments/non-evm-deployments/solana
+npm install
+npm run setup
+```
+
+**After deployment, proceed to [Usage](#usage) to start creating invoices.**
+
+For detailed deployment steps, continue below.
+
+## Deployment
+
+### 1. Install Dependencies
+
````
think we have to point out multi region deployments - consider using the `env` file as single source of truth here and propagate through the deployment process (think we are short on this on the `evm` stack as well)

racket2000 replied:
DONE -- incorp'd region handling into SOL Blueprint 

I added note to run AWS Configure for manual deployment, along with a AWS_REGION Variable that gets written into .env during the `setup` script

dpdornseifer reviewed (changes requested):
lgtm - pointed out 4 minor issues

Commit:
upgrade from js sdk v2 to v3, add region var, enable cdknag, remove hard coded log group name from watcher, remove kms key by default

Commit:
fix formatting

Commit:
fix sweeper bug regarding rent

Commit:
Remove unnecessary top-up for SPL sweeps since hot wallet pays all fees

racket2000 commented:
Addressed above feedback, and refactored JS SDK (used in 4 Lambda's) from v2 to v3

This should:
1. Reduce package size (This should ultimately cut down on CloudShell out of storage errors) 
2. Reduce latency (especially on cold starts)

@dpdornseifer , can you please take another look at your own convenience?

Commit:
resolve linter findings

Commit:
update package-lock.json

Commit:
suppress bigint warning in SPL payment test script

Commit:
filter bigint warning while preserving error messages

Commit:
replace usdc mainnet address with testnet address

racket2000 commented:
final testing and validation complete - merging to main
